cmd/format/format.go

```diff
@@ -205,33 +205,34 @@ func Run(v *viper.Viper, statz *stats.Stats, cmd *cobra.Command, paths []string)
 		return fmt.Errorf("failed to create walker: %w", err)
 	}
 
+	// start traversing
 	files := make([]*walk.File, BatchSize)
+
 	for {
+		// read the next batch
 		ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
-
 		n, err := reader.Read(ctx, files)
 
-		for idx := 0; idx < n; idx++ {
-			file := files[idx]
+		// ensure context is cancelled to release resources
+		cancel()
 
-			// check if this file is new or has changed when compared to the cache entry
-			if file.Cache == nil || file.Cache.HasChanged(file.Info) {
-				filesCh <- file
-				statz.Add(stats.Emitted, 1)
-			}
+		// pass each file into the file channel for processing
+		for idx := 0; idx < n; idx++ {
```

Comment:
The `applyFormatters` goroutine has state:

- an `errgroup.Group` limited to `runtime.NumCPU()` for executing the formatting tasks
- a map of `batch key -> []*format.Task` which is used to group formatting tasks by the unique ordered set of formatters to be applied. 

We feed it files via the channel, it assigns them to the correct batch based on the batch key, and then triggers a format if the batch reaches the `BatchSize`, flushing any partial batches when the file channel is closed. 

We want the above logic to be happening concurrently with traversing the file system, invoking formatters as soon as we can. The batch size is set to `1024` at the moment, so most of the time the formatters are getting invoked after we close the channel. But for larger repos it will have a pipeline effect. 

There's also #335 (which needs rebased) to allow for finer grained control of the batch size.

Reply:
Ah! This really helps me understand the architecture, thank you.

Reply:
Also, for a little more context, when I wrote this comment this morning, I (incorrectly) thought that checking if a file is cached is a moderately expensive operation (I was thinking we had to read the whole file to compute a hash). I have since read the relevant cache query code and I see that it's just based on time + filesize.

Reply:
Yeah, we may eventually make the comparison strategy configurable and allow users to decide how expensive they want it to be. But since its inception, this is the heuristic `treefmt` has used. 

There is an issue that discusses the pros and cons of different approaches; I'll try to find it.